docs/tsm/TSI_CARDINALITY_PROPOSAL.md

```diff
@@ -0,0 +1,254 @@
+The reason being, that any series tombstoned in the index or WAL will also have been accounted for and if we don't consider them we will double-count them, and reduce the accuracy of the cardinality estimate.
+Possibly we could avoid this and have a less accurate estimation.
+Thoughts would be welcome on the above...
+
+
+##### Fast compactions
+
+During a fast compaction, WAL data is merged into the associated TSI index.
+We can combine the WAL and TSI sketches during this process, persisting the updated sketches to the TSI file.
+We would clear the in-memory tombstone counts, and reload them from the new tombstone counts in the TSI indexes.
+
+##### Full compactions
+
+During a full compaction, tombstoned series and measurements are removed from the TSI indices.
+At this point a new sketches will need to be calculated for the index, which should replace any existing ones.
+These new sketches would replace the current in-memory sketches.
+
+#### TSI file format
+
+We could extend the TSI file format to include a sketches block for series and measurements, which could be for example at the beginning of the index.
+We would also need to add tombstone counters for series and measurements, which I have not added to this diagram, but would need to be considered in #7174.
+
+```
+╔═══════Inverted Index═══════╗
+║ ┌────────────────────────┐ ║
+║ │                        │ ║
+║ │        Sketches        │ ║
+║ │                        │ ║
+║ └────────────────────────┘ ║
+║ ┌────────────────────────┐ ║
+║ │                        │ ║
+║ │   Series Dictionary    │ ║
+║ │                        │ ║
+║ └────────────────────────┘ ║
+║ ┌────────Tag Set─────────┐ ║
+║ │┌──────────────────────┐│ ║
+║ ││   Tag Values Block   ││ ║
+║ │├──────────────────────┤│ ║
+║ ││   Tag Values Block   ││ ║
+║ │├──────────────────────┤│ ║
+║ ││    Tag Keys Block    ││ ║
+║ │└──────────────────────┘│ ║
+║ └────────────────────────┘ ║
+║ ┌────────Tag Set─────────┐ ║
+║ │┌──────────────────────┐│ ║
+║ ││   Tag Values Block   ││ ║
+║ │├──────────────────────┤│ ║
+║ ││   Tag Values Block   ││ ║
+║ │├──────────────────────┤│ ║
+║ ││   Tag Values Block   ││ ║
+║ │├──────────────────────┤│ ║
+║ ││    Tag Keys Block    ││ ║
+║ │└──────────────────────┘│ ║
+║ └────────────────────────┘ ║
+║  ┌──────────────────────┐  ║
+║  │  Measurements Block  │  ║
+║  └──────────────────────┘  ║
+╚════════════════════════════╝
+```
+
+The sketch block itself would contain a value indicating the number of sketches in the block.
+Each sketch block would then contain the size of the sketch data, and the sketch itself.
+The sketch may be a binary blob, perhaps generated by Go's `gob` package, or it may be a custom binary format of data comprising the HLL algorithm state and sketch.
+The exact format will be decided during implementation.
+
+```
+╔═══════════Sketches═══════════╗
+║ ┌──────────────────────────┐ ║
+║ │   Sketch Count <uint32>  │ ║
```

Comment:
Since sketches are a fixed size we would then know how far to seek, to get to the dictionary if we added further sketch-types to the Sketches block in the future. Since @benbjohnson pointed out that the Sketch block should be after the dictionary anyway, then maybe we don't need this?